Pipeline decorator: Add option not to modify the doc string

lschr opened the pull request:
The `pipeline` decorator adds a note into the wrapped function's doc string saying that this function has been made lazy. This however leads to breakage with sphinx autodoc, since in general the indentation level of what is added does not match that of the doc string.

Since it would be rather hard (I think) to match indentation levels, I propose adding a keyword argument to the decorator which will cause it not to modify the doc string at all.

With this PR, the following constructs will lead to the old behavior (i. e. doc string with the note added)

``` python
@pipeline
def f(x):
    """Doc string"""
    pass

@pipeline()
def f(x):
    """Doc string"""
    pass

@pipeline(retain_doc=False)
def f(x):
    """Doc string"""
    pass
```

while the following will leave the doc string as is (no note added):

``` python
@pipeline(retain_doc=True)
def f(x):
    """Doc string"""
    pass
```

Commit:
Add `retain_doc` argument for pipeline decorator

When called with retain_doc=True, the pipeline decorator will not
alter the original doc string. This is useful especially in conjunction
with sphinx autodoc since changing the doc string messes up indentation,
which leads to errors.

caspervdw commented on slicerator.py:
```diff
@@ -493,9 +506,35 @@ def pipeline(func):
     >>> red_img = red_channel(single_img)  # normal behavior
     """
+    def wrapper(f):
+        return _pipeline(f, kwargs.get("retain_doc", False))
```
Why not pass all `**kwargs` over here? Now typos like `retin_doc` will be silently accepted.

caspervdw commented:
:+1: thanks @lschr , I left one comment, but for the rest it can be merged.
The pattern for the wrapper makes room for additional `kwargs` that tune attribute propagation through the pipeline #16, which is nice!

Commit:
Pass all kwargs from `pipeline` to `_pipeline`

This a) errors out on invalid keywords, typos, etc. and b) makes it
easier to add additional keyword args to `_pipeline`; they just need to
be added in the `_pipeline` def.

lschr commented:
@caspervdw I hope that was what you suggested. Anyways, it is much nicer now (in my opinion).

tacaswell commented:
https://docs.python.org/3.5/library/textwrap.html has some tools for managing indentation.  mpl has a fair amount of this munging and it is a pain to carry around, but this decorator changes the semantics of the function enough it might be worth the effort here.

I am :+1: on this change either way.

lschr commented:
The problem with `textwrap.dedent` and `matplotlib.cbook.dedent` is the fact that when creating a doc string

``` python
def func():
    """Doc string

    with multiple
    lines
    """
```

the result is

```
Doc string

    with multiple
    lines
```

i. e. the first line is not indented and therefore the unindent functions don't work. In matplotlib the work-around apparently is to define the docstrings with a leading empty line,

``` python
def func():
    """
    Doc string

    with multiple
    lines
    """
```

So, I guess our options are
- Have the users of `pipeline` do the same (empty first line) and include something like `matplotlib.cbook.dedent` here.
- Ship a modified version of `matplotlib.cbook.dedent` that gets the indentation from another line but the first. This may not be so easy since there probably are some corner cases that need to be accounted for as well.
- Leave the doc string as is and have the user add a note about the pipeline

caspervdw commented:
Thanks @lschr for diving into this doc-adaptation. I think it would be nice to give this some more thought. What do we want exactly? Personally, I would not like to prepend the `__doc__` but add it more after the first sentence. Possibly also something saying what the pipeline actually does.

I think this is something for a different time though, as it is this PR is already fine. So I leave it as an open issue for now.
